src/DataCollectors/TraceDataCollector/VanguardCollector/DynamicCoverageDataCollectorImpl.cs

```diff
@@ -0,0 +1,365 @@
+        /// Logger
+        /// </summary>
+        private IDataCollectionLogger logger;
+
+        /// <summary>
+        /// Data sink
+        /// </summary>
+        private TraceCollector.IDataCollectionSink dataSink;
+
+        /// <summary>
+        /// Directory helper
+        /// </summary>
+        private IDirectoryHelper directoryHelper;
+
+        /// <summary>
+        /// File helper
+        /// </summary>
+        private IFileHelper fileHelper;
+
+        /// <summary>
+        /// Folder to store temporary files
+        /// </summary>
+        private string sessionDirectory;
+
+        public DynamicCoverageDataCollectorImpl()
+        : this(new Vanguard(), new DirectoryHelper(), new FileHelper())
+        {
+        }
+
+        internal DynamicCoverageDataCollectorImpl(IVangurd vanguard, IDirectoryHelper directoryHelper, IFileHelper fileHelper)
+        {
+            this.Vanguard = vanguard;
+            this.directoryHelper = directoryHelper;
+            this.fileHelper = fileHelper;
+        }
+
+        /// <summary>
+        /// Gets or sets gets session name
+        /// </summary>
+        private string SessionName { get; set; }
+
+        /// <summary>
+        /// Gets or sets vanguard instance
+        /// </summary>
+        private IVangurd Vanguard { get; set; }
+
+        public string GetSessionName()
+        {
+            return this.SessionName;
+        }
+
+        /// <summary>
+        /// Initialize
+        /// </summary>
+        /// <param name="configurationElement">Configuration element</param>
+        /// <param name="dataSink">Data sink</param>
+        /// <param name="logger">Logger</param>
+        public virtual void Initialize(
+            XmlElement configurationElement,
+            TraceCollector.IDataCollectionSink dataSink,
+            IDataCollectionLogger logger)
+        {
+            EqtTrace.Info("DynamicCoverageDataCollectorImpl.Initialize: Initialize configuration. ");
+            if (string.IsNullOrEmpty(configurationElement?.InnerXml))
+            {
+                // Add default configuration specific to CodeCoverage.
+                var doc = new XmlDocument();
+                using (
+                    var xmlReader = XmlReader.Create(
+                        new StringReader(DefaultCodeCoverageSettings),
+                        new XmlReaderSettings() { CloseInput = true, DtdProcessing = DtdProcessing.Prohibit }))
+                {
+                    doc.Load(xmlReader);
+                }
+
+                configurationElement = doc.DocumentElement;
+            }
+
+            this.logger = logger;
+            this.dataSink = dataSink;
+
+            this.dataSink.SendFileCompleted += this.OnSendFileCompletedEvent;
+
+            this.SessionName = Guid.NewGuid().ToString();
+
+            this.sessionDirectory = Path.Combine(Path.GetTempPath(), this.SessionName);
+            this.directoryHelper.CreateDirectory(this.sessionDirectory);
```

Comment:
This can fail, please put try-catch, or make it TryCreateDirectory, same for DeleteDirectory

Reply:
Done.